pkg/resource/graph/dependency_graph_test.go

```diff
@@ -75,841 +82,1680 @@ func TestDependencyGraph(t *testing.T) {
+				a1,        // a2 depends on a1
+			)
+
+			// Act.
+			actual := dg.DependenciesOf(a2)
+
+			// Assert.
+			assertSameSets(t, expected, actual)
+		})
+
+		t.Run("providerB", func(t *testing.T) {
+			t.Parallel()
+
+			// Arrange.
+			expected := mapset.NewSet(
+				a1, // providerB depends on a1
+				a2, // providerB depends on a2
+			)
+
+			// Act.
+			actual := dg.DependenciesOf(providerB)
+
+			// Assert.
+			assertSameSets(t, expected, actual)
+		})
+
+		t.Run("b1", func(t *testing.T) {
+			t.Parallel()
+
+			// Arrange.
+			expected := mapset.NewSet(
+				providerB, // b1's provider is providerB
+				a1,        // b1 depends on a1
+			)
+
+			// Act.
+			actual := dg.DependenciesOf(b1)
+
+			// Assert.
+			assertSameSets(t, expected, actual)
+		})
+
+		t.Run("c1", func(t *testing.T) {
+			t.Parallel()
+
+			// Arrange.
+			expected := mapset.NewSet(
+				a2, // c1 depends on a2
+			)
+
+			// Act.
+			actual := dg.DependenciesOf(c1)
+
+			// Assert.
+			assertSameSets(t, expected, actual)
+		})
+
+		t.Run("providerD", func(t *testing.T) {
+			t.Parallel()
+
+			// Arrange.
+			expected := mapset.NewSet[*resource.State]()
+
+			// Act.
+			actual := dg.DependenciesOf(providerD)
+
+			// Assert.
+			assertSameSets(t, expected, actual)
+		})
+
+		t.Run("d1", func(t *testing.T) {
+			t.Parallel()
+
+			// Arrange.
+			expected := mapset.NewSet(
+				providerD, // d1's provider is providerD
+			)
+
+			// Act.
+			actual := dg.DependenciesOf(d1)
+
+			// Assert.
+			assertSameSets(t, expected, actual)
+		})
+
+		t.Run("d2", func(t *testing.T) {
+			t.Parallel()
+
+			// Arrange.
+			expected := mapset.NewSet(
+				d1, // d2 is a child of d1
+			)
+
+			// Act.
+			actual := dg.DependenciesOf(d2)
+
+			// Assert.
+			assertSameSets(t, expected, actual)
+		})
+
+		t.Run("d3", func(t *testing.T) {
+			t.Parallel()
+
+			// Arrange.
+			expected := mapset.NewSet(
+				providerD, // d3's provider is providerD
+				d1,        // d3 is deleted with d1 (a component)
+				d2,        // d2 is a child of d1 that appears before d3
+			)
+
+			// Act.
+			actual := dg.DependenciesOf(d3)
+
+			// Assert.
+			assertSameSets(t, expected, actual)
+		})
+
+		t.Run("d6", func(t *testing.T) {
+			t.Parallel()
+
+			// Arrange.
+			expected := mapset.NewSet[*resource.State]()
+
+			// Act.
+			actual := dg.DependenciesOf(d6)
+
+			// Assert.
+			assertSameSets(t, expected, actual)
+		})
+
+		t.Run("d7", func(t *testing.T) {
+			t.Parallel()
+
+			// Arrange.
+			expected := mapset.NewSet(
+				d6, // d7 is a child of d6
+			)
+
+			// Act.
+			actual := dg.DependenciesOf(d7)
+
+			// Assert.
+			assertSameSets(t, expected, actual)
+		})
+
+		t.Run("d8", func(t *testing.T) {
+			t.Parallel()
+
+			// Arrange.
+			expected := mapset.NewSet(
+				d6, // d8 is deleted with d6
+				// d7 does not appear as a child of d6 because d6 is not a component
+			)
+
+			// Act.
+			actual := dg.DependenciesOf(d8)
+
+			// Assert.
+			assertSameSets(t, expected, actual)
+		})
```

Comment:
The previous three tests are new and test the behaviour when a non-component (custom, `Custom: true`) resource has children.